File: 2025/JuyoungMoon/week11/2025-08-17-longest-increasing-subsequence.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 11: DP 2
// https://github.com/juyomo/leetcode-study

// LeetCode #300.
// https://leetcode.com/problems/longest-increasing-subsequence/

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        int max = 1;
        for (int i = 0; i < dp.size(); i++) {
            if (dp[i] > max) {
                max = dp[i];
            }
        }
        return max;
    }
};

File: 2025/jimmycho/week11/2025-08-17-Longest-Increasing-Subsequence.py
class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        res = 1
        dp = [[1] * (len(nums) + 1) for _ in range(len(nums) + 1)]
        for i in range(1, len(nums) + 1):
            dp[i][i] = dp[i - 1][i]
            for j in range(i + 1, len(nums) + 1):
                if nums[i - 1] < nums[j - 1]:
                    dp[i][j] = max(dp[i][i] + 1, dp[i - 1][j])
                else:
                    dp[i][j] = dp[i - 1][j]
                res = max(res, dp[i][j])
        return res
